### PR TITLE
fix(math): correct xlogy/xlog1py gradient w.r.t. x at x=0

### LINKED ISSUE #119476: `tf.math.xlogy` and `tf.math.xlog1py` gradient w.r.t. x is incorrectly zero when x=0
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

 2.22.0-dev20260508

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Summary

`tf.math.xlogy(x, y)` and `tf.math.xlog1py(x, y)` return incorrect gradient `0` w.r.t. `x` when `x=0` and `y > 0`. The correct gradients are `log(y)` and `log(1+y)` respectively. PyTorch's `torch.xlogy` correctly returns `log(y)` in this case.

### PyTorch comparison

PyTorch correctly returns the gradient at x=0:

```python
import torch
x = torch.tensor(0.0, requires_grad=True, dtype=torch.float64)
y = torch.tensor(2.0, dtype=torch.float64)
torch.xlogy(x, y).backward()
print(x.grad)  # tensor(0.6931, dtype=torch.float64)  -- correct: log(2)
```

### Root cause

Both functions are defined piecewise: `xlogy(0, y) = 0` to handle `0 * log(0) = 0`. The gradient w.r.t. `x` is `log(y)` for all `x` including `x = 0` (when `y > 0`):

```
d/dx xlogy(x, y)|_{x=0} = lim_{h→0} [h·log(y) - 0] / h = log(y)
```

The implementation applies a zero-mask (from the `x == 0` special case in the forward pass) to the gradient as well, but should only apply it to the function value, not the derivative w.r.t. `x`. The gradient w.r.t. `y` is unaffected (correctly returns `x/y = 0` when `x = 0`).

### Impact

This creates a dead zone where the optimizer receives zero gradient and cannot update the parameter through zero. Affects:
- KL divergence computations where class probabilities are zero
- Cross-entropy losses with zero-weighted components  
- Mixture models where component weights pass through zero during optimization

### Environment

- TensorFlow: 2.22.0-dev20260508
- OS: Ubuntu 20.04
- Affects both CPU and GPU

### Standalone code to reproduce the issue

```shell
### Reproduction


import tensorflow as tf

# xlogy
x = tf.constant([0.0, 0.0, 1.0], dtype=tf.float64)
y = tf.constant([2.0, 5.0, 2.0], dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x)
    out = tf.math.xlogy(x, y)
g = tape.gradient(out, x)
print("TF xlogy grad:", g.numpy())   # [0.         0.         0.69314718]
# Correct:                            # [0.69314718 1.60943791 0.69314718]

# xlog1py
x2 = tf.constant([0.0, 0.0, 1.0], dtype=tf.float64)
y2 = tf.constant([1.0, 4.0, 1.0], dtype=tf.float64)

with tf.GradientTape() as tape:
    tape.watch(x2)
    out2 = tf.math.xlog1py(x2, y2)
g2 = tape.gradient(out2, x2)
print("TF xlog1py grad:", g2.numpy())  # [0.         0.         0.69314718]
# Correct:                              # [0.69314718 1.60943791 0.69314718]
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by goingforstudying-ctrl**
## Summary
This PR fixes the gradient computation for `tf.math.xlogy` and `tf.math.xlog1py` with respect to `x` when `x = 0`.

**Bug:** When `x = 0` and `y > 0`, the gradient w.r.t. `x` was incorrectly returning `0` instead of `log(y)` (for `xlogy`) or `log1p(y)` (for `xlog1py`).

**Root cause:** The gradient implementation reused the same zero-mask (`not_zero_x`) that is correct for the *forward* pass (to define `xlogy(0, y) = 0`), but incorrectly applied it to the *gradient* w.r.t. `x`. The derivative `d/dx x·log(y)` is `log(y)` for all `x`, including `x = 0`, because it is a linear function of `x`.

**Impact:** This bug created dead zones in optimization where parameters passing through zero received no gradient, affecting KL divergence, cross-entropy, and mixture model training.

## Changes
- `tensorflow/python/ops/math_grad.py`
  - `_XLogyGrad`: replaced `gen_math_ops.xlogy(not_zero_x, y)` with `gen_math_ops.log(y)`
  - `_XLog1pyGrad`: replaced `gen_math_ops.xlog1py(not_zero_x, y)` with `gen_math_ops.log1p(y)`
- `tensorflow/python/ops/math_grad_test.py`
  - Updated `XlogyTest.testZeroXGrad` to expect `log(y)` instead of `0`
  - Updated `Xlog1pyTest.testZeroXGrad` to expect `log1p(y)` instead of `0`

## Test Plan
- `bazel test //tensorflow/python/ops:math_grad_test --test_filter="XlogyTest|Xlog1pyTest"`
- Verified the reproduction script from #119476 now returns the correct gradients matching PyTorch behavior.

Fixes #119476

### CODE REVIEW

**Comment by goingforstudying-ctrl** [comment]
@gemini-code-assist Thanks for the review suggestions! I have added the edge-case test coverage in the latest push:

1. `testZeroXGrad` for `xlogy` now also tests `y = 0.0` to verify the gradient w.r.t. `x` evaluates correctly at that boundary.
2. `testZeroXGrad` for `xlog1py` now also tests `y = -1.0` to verify the gradient w.r.t. `x` evaluates correctly when `1 + y = 0`.

Please let me know if any additional test cases are needed.

**Comment by keerthanakadiri** [comment]
Hi @goingforstudying-ctrl , Can you kindly sign the CLA? Thank you!

**Comment by goingforstudying-ctrl** [comment]
> Hi @goingforstudying-ctrl , Can you kindly sign the CLA? Thank you!

HI, already signed. Please take a look when you have chance. Thanks!

**Comment by goingforstudying-ctrl** [comment]
Hi keerthanakadiri, I have already signed the Google CLA. The check may need a re-trigger. Could you please advise if there is anything else I need to do on my end?

**Comment by goingforstudying-ctrl** [comment]
Thanks for the review! The edge case tests for `y = 0.0` (xlogy) and `y = -1.0` (xlog1py) are already included in the latest commit:

- `testZeroXGrad` now loops over `[3.1, 0.0]` for xlogy and `[3.1, -1.0]` for xlog1py
- `testZeroXYGrad` and `testZeroXNegOneYGrad` assert `-np.inf` for the gradient w.r.t. x at the boundary conditions

These should cover the boundary behavior you mentioned.

**Comment by goingforstudying-ctrl** [comment]
Hi, I reviewed the gemini-code-assist suggestions about adding boundary tests.

For xlogy: the x=0, y=0 boundary is already covered by testZeroXYGrad (line 547-555), which asserts the x gradient is -inf and the y gradient is 0 at that point.

For xlog1py: the x=0, y=-1 boundary is already covered by testZeroXNegOneYGrad (line 601-610), which asserts the x gradient is -inf and the y gradient is 0 at that point.

Both edge cases are tested in dedicated tests rather than inside testZeroXGrad, so the coverage is complete.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for submitting this correction! You are absolutely right: the partial derivative of $x \log(y)$ with respect to $x$ is indeed $\log(y)$, even at $x=0$. Applying the forward pass zero-expression-mask to the backward pass was a calculation error that incorrectly muted gradients at the origin.
  
  To merge this, we need to update the corresponding unit tests to reflect the new, mathematically correct expectations (including the $-\infty$ evaluations at boundary conditions), otherwise the automated test checks will fail.
  
  Please update the following assertions in `tensorflow/python/ops/math_grad_test.py` in your PR:
  
  1. In `XlogyTest.testZeroXGrad` (around line 530), replace:
     ```python
     self.assertAllClose(zero, xlogy_xgrad)
     ```
     with:
     ```python
     self.assertAllClose(self.evaluate(math_ops.log(y)), xlogy_xgrad)
     ```
  
  2. In `XlogyTest.testZeroXYGrad` (around line 549), replace:
     ```python
     self.assertAllClose(zero, xlogy_xgrad)
     ```
     with:
     ```python
     self.assertAllClose(-np.inf, xlogy_xgrad)
     ```
  
  3. In `Xlog1pyTest.testZeroXGrad` (around line 580), replace:
     ```python
     self.assertAllClose(zero, xlog1py_xgrad)
     ```
     with:
     ```python
     self.assertAllClose(self.evaluate(math_ops.log1p(y)), xlog1py_xgrad)
     ```
  
  4. In `Xlog1pyTest.testZeroXNegOneYGrad` (around line 599), replace:
     ```python
     self.assertAllClose(zero, xlog1py_xgrad)
     ```
     with:
     ```python
     self.assertAllClose(-np.inf, xlog1py_xgrad)
     ```
  
  Once those unit tests are updated and PR checks are green, we will proceed with the approval and merge!

**Review by dmiltr3 — DISMISSED** [review]

**Review by dmiltr3 — DISMISSED** [review]

**Comment by goingforstudying-ctrl** [comment]
Thanks for the approval, @dmiltr3. The public checks and CLA are green now. The only remaining red status I see is `feedback/copybara`, which looks like an internal Copybara feedback check. Could a maintainer please take a look or advise if anything is still needed from my side?

**Review by dmiltr3 — DISMISSED** [review]

**Review by dmiltr3 — APPROVED** [review]

**Comment by dmiltr3** [comment]
Hi, since changes are approved, we do not expect any new commits like [Merge branch 'master' into fix-xlogy-gradient-zero-x](https://github.com/tensorflow/tensorflow/pull/119869/commits/ec497af7d3cdafd18ca822e1e632aba863813ad4) even empty ones, otherwise the code synchronization engine is getting confused and it stops the code migration process.

**Comment by keerthanakadiri** [comment]
> Hi, since changes are approved, we do not expect any new commits like [Merge branch 'master' into fix-xlogy-gradient-zero-x](https://github.com/tensorflow/tensorflow/pull/119869/commits/ec497af7d3cdafd18ca822e1e632aba863813ad4) even empty ones, otherwise the code synchronization engine is getting confused and it stops the code migration process.

Hi @goingforstudying-ctrl, please check this.

**Comment by goingforstudying-ctrl** [comment]
Thanks for flagging this. I understand now and will not push any further commits or merge updates unless requested by maintainers. I see the Copybara checks are currently green; please let me know if anything else is needed from my side.

**Review by dmiltr3 — APPROVED** [review]
